### ISSUE TITLE
Compiler errors on latest git pull

### ISSUE THREAD

**Issue by elimpnick** [stealth/opmsg#15]
Hi,

I am getting a compiler error building as follows:

c++ -Wall -O2 -pedantic -std=c++11   -c keystore.cc
keystore.cc: In member function ‘int opmsg::keystore::load(const string&, uint32_t)’:
keystore.cc:201:7: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
   if (readdir_r(d, &de, &result) < 0)
       ^~~~~~~~~
In file included from keystore.cc:29:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
keystore.cc:201:32: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
   if (readdir_r(d, &de, &result) < 0)
                                ^
In file included from keystore.cc:29:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
keystore.cc: In member function ‘int opmsg::persona::load(const string&, uint32_t)’:
keystore.cc:852:7: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
   if (readdir_r(d, &de, &result) < 0)
       ^~~~~~~~~
In file included from keystore.cc:29:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
keystore.cc:852:32: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
   if (readdir_r(d, &de, &result) < 0)
                                ^
In file included from keystore.cc:29:0:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
keystore.cc: In member function ‘int opmsg::persona::gen_dh_key(const EVP_MD*, std::__cxx11::string&, std::__cxx11::string&, std::__cxx11::string&)’:
keystore.cc:1152:24: error: invalid conversion from ‘BIGNUM** {aka bignum_st**}’ to ‘const BIGNUM** {aka const bignum_st**}’ [-fpermissive]
  DH_get0_key(dh.get(), &pub_key, nullptr);
                        ^~~~~~~~
In file included from keystore.cc:35:0:
/usr/include/openssl/dh.h:158:6: note:   initializing argument 2 of ‘void DH_get0_key(const DH*, const BIGNUM**, const BIGNUM**)’
 void DH_get0_key(const DH *dh,
      ^~~~~~~~~~~
keystore.cc: In member function ‘std::vector<opmsg::PKEYbox*> opmsg::persona::add_dh_pubkey(const EVP_MD*, std::vector<std::__cxx11::basic_string<char> >&)’:
keystore.cc:1247:26: error: invalid conversion from ‘BIGNUM** {aka bignum_st**}’ to ‘const BIGNUM** {aka const bignum_st**}’ [-fpermissive]
    DH_get0_key(dh.get(), &pub_key, nullptr);
                          ^~~~~~~~
In file included from keystore.cc:35:0:
/usr/include/openssl/dh.h:158:6: note:   initializing argument 2 of ‘void DH_get0_key(const DH*, const BIGNUM**, const BIGNUM**)’
 void DH_get0_key(const DH *dh,
      ^~~~~~~~~~~
make: *** [Makefile:91: keystore.o] Error 1

This is with gcc (GCC) 6.3.1 20170306.  Any ideas?

**Comment by stealth** [comment]
Its not about the compiler. Which distro and openssl install were you using?
I get similar result on a Fedora26 Alpha.
The problem is that openssl again changed and broke their API by declaring
DH_get0_key() parameters to const.
So I had to wrap it. The patch is already in the git. Just pull and make.
readdir_r() has not yet been addresses, as its just a warning.
For whatever reason POSIX choose to deprecate the thread safe version
vs. the unsafe version -.-
Closing as fixed
